File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1;
        int best = 0;
        while (l < r){
            int curr = (r - l) * min(heights[l], heights[r]);
            best = max(best, curr);
            if (heights[l] < heights[r]){
                l++;
            }
            else{
                r--;
            }
        }
        return best;
    }
};

File: Data Structures & Algorithms/insert-new-interval/submission-4.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # iterate through until curr start is after new 
        # if start or end is found within curr interval, take the min(start) and max(end) -> update "newInterval"
        #    keep stepping forward until end is BEFORE next, then we can add it to our result.. 
        res = []

        for i in range(len(intervals)):
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1]),
                ]
        res.append(newInterval)
        return res
